browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+      url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  // GetDomainAndRegistry might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.empty())
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
+
+  return domain;
+}
+
+}  // namespace
+namespace ephemeral_storage {
+
+EphemeralStorageTabHelper::~EphemeralStorageTabHelper() {}
+
+EphemeralStorageTabHelper::EphemeralStorageTabHelper(WebContents* web_contents)
+    : WebContentsObserver(web_contents) {}
+
+void EphemeralStorageTabHelper::ReadyToCommitNavigation(
+    NavigationHandle* navigation_handle) {
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage))
+    return;
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  std::string domain = URLToStorageDomain(navigation_handle->GetURL());
+  std::string previous_domain =
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (domain == previous_domain)
+    return;
+
+  ClearEphemeralStorageIfNecessary();
+
+  auto* browser_context = web_contents()->GetBrowserContext();
+  auto site_instance = content::SiteInstance::CreateForURL(
+      browser_context, navigation_handle->GetURL());
+  auto* partition =
+      BrowserContext::GetStoragePartition(browser_context, site_instance.get());
+
+  std::string local_partition_id = domain + "/ephemeral-local-storage";
+  content::SessionStorageNamespaceMap::const_iterator it =
+      local_storage_namespace_map().find(local_partition_id);
+  if (it == local_storage_namespace_map().end()) {
+    auto local_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, local_partition_id);
+    local_storage_namespace_map()[local_partition_id] =
+        std::move(local_storage_namespace);
+  }
+
+  std::string session_partition_id =
+      content::GetSessionStorageNamespaceId(web_contents()) +
+      "/ephemeral-session-storage";
+  it = session_storage_namespace_map().find(session_partition_id);
+  if (it == session_storage_namespace_map().end()) {
+    auto session_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, session_partition_id);
+    session_storage_namespace_map()[session_partition_id] =
+        std::move(session_storage_namespace);
+  }
+}
+
+void EphemeralStorageTabHelper::WebContentsDestroyed() {
+  ClearEphemeralStorageIfNecessary();
+}
+
+bool EphemeralStorageTabHelper::IsDifferentWebContentsWithMatchingStorageDomain(
+    WebContents* other_web_contents,
+    const std::string& storage_domain) {
+  DCHECK(other_web_contents);
+  if (web_contents() == other_web_contents)
+    return false;
+
+  std::string other_storage_domain =
+      URLToStorageDomain(other_web_contents->GetLastCommittedURL());
+  return storage_domain == other_storage_domain;
+}
+
+bool EphemeralStorageTabHelper::IsAnotherTabOpenWithStorageDomain(
+    const std::string& storage_domain) {
```

Comment:
`IsAnotherTabOpenWithStorageDomain` probably no need to pass `storage_domain` if this is always `URLToStorageDomain(web_contents()->GetLastCommittedURL())`

Reply:
This is to avoid converting the previously committed URL to a storage domain two times.